File: src/hooks/parse.rs
//! Parse hook-config files into the normalized [`NormalizedHook`] IR.
//!
//! These are **input-format** parsers for the shared `collect` pipeline — not per-harness logic.
//! `collect` picks a parser by the *source format* of the file, not by the harness being targeted:
//! every pack plugin / bare skill / `.agents` tree authors hooks in the canonical nested format
//! ([`parse_nested_hooks`]) and is parsed once, then fanned out to all harness renderers; only the
//! user's seeded native Codex `hooks.json` uses the Codex CLI's own format ([`parse_codex_hooks`]).

use std::collections::BTreeMap;
use std::io::{self, ErrorKind};
use std::path::Path;

use serde_json::{Map, Value};

use crate::error::{AgentpackError, Result};

use super::ir::{
    AgentHandler, ClaudeEvent, ClaudeHandler, CommandHandler, HookBundle, HookOrigin, HttpHandler,
    NormalizedHook, PromptHandler,
};

/// Parse the canonical nested hook format used by all pack content and `.agents`:
/// `{ "hooks": { <event>: [ { "matcher"?, "hooks": [ <handler>, … ] }, … ] } }`.
pub fn parse_nested_hooks(
    path: &Path,
    value: &Value,
    base_origin: &HookOrigin,
) -> Result<HookBundle> {
    let hooks_object = value
        .get("hooks")
        .and_then(Value::as_object)
        .ok_or_else(|| invalid(path, "hook file must contain a top-level hooks object"))?;

    let mut hooks = Vec::new();
    for (event_index, (event_name, groups)) in hooks_object.iter().enumerate() {
        let event = ClaudeEvent::from_any_str(event_name)
            .ok_or_else(|| invalid(path, format!("unknown hook event `{event_name}`")))?;
        let groups = groups
            .as_array()
            .ok_or_else(|| invalid(path, "event groups must be arrays"))?;
        for (matcher_group_index, group) in groups.iter().enumerate() {
            let group_obj = group
                .as_object()
                .ok_or_else(|| invalid(path, "matcher groups must be objects"))?;
            let matcher = group_obj
                .get("matcher")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned);
            let group_extra = object_extra(group_obj, &["matcher", "hooks"]);
            let group_hooks = group_obj
                .get("hooks")
                .and_then(Value::as_array)
                .ok_or_else(|| invalid(path, "matcher groups must contain hooks arrays"))?;
            for (hook_index, handler_value) in group_hooks.iter().enumerate() {
                let mut origin = base_origin.clone();
                origin.event_index = event_index;
                origin.matcher_group_index = matcher_group_index;
                origin.hook_index = hook_index;
                hooks.push(build_hook(
                    event,
                    matcher.clone(),
                    handler_value,
                    origin,
                    group_extra.clone(),
                    path,
                )?);
            }
        }
    }

    Ok(HookBundle { hooks })
}

/// Parse the Codex CLI's native `hooks.json` (the seeded user config). Like the nested format but
/// also accepts entries that are handlers directly (no inner `hooks` array).
pub fn parse_codex_hooks(
    path: &Path,
    value: &Value,
    base_origin: &HookOrigin,
) -> Result<HookBundle> {
    let hooks_root = if let Some(hooks) = value.get("hooks") {
        hooks
    } else {
        value
    };
    let hooks_object = hooks_root
        .as_object()
        .ok_or_else(|| invalid(path, "Codex hooks file must be an object"))?;

    let mut hooks = Vec::new();
    for (event_index, (event_name, entries)) in hooks_object.iter().enumerate() {
        let event = ClaudeEvent::from_any_str(event_name)
            .ok_or_else(|| invalid(path, format!("unknown Codex hook event `{event_name}`")))?;
        let entries = entries
            .as_array()
            .ok_or_else(|| invalid(path, "Codex hook event entries must be arrays"))?;
        for (matcher_group_index, entry) in entries.iter().enumerate() {
            let entry_obj = entry
                .as_object()
                .ok_or_else(|| invalid(path, "Codex hook entries must be objects"))?;
            if let Some(group_hooks) = entry_obj.get("hooks").and_then(Value::as_array) {
                let matcher = entry_obj
                    .get("matcher")
                    .and_then(Value::as_str)
                    .map(ToOwned::to_owned);
                let group_extra = object_extra(entry_obj, &["matcher", "hooks"]);
                for (hook_index, handler_value) in group_hooks.iter().enumerate() {
                    let mut origin = base_origin.clone();
                    origin.event_index = event_index;
                    origin.matcher_group_index = matcher_group_index;
                    origin.hook_index = hook_index;
                    hooks.push(build_hook(
                        event,
                        matcher.clone(),
                        handler_value,
                        origin,
                        group_extra.clone(),
                        path,
                    )?);
                }
                continue;
            }

            let matcher = entry_obj
                .get("matcher")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned);
            let group_extra = object_extra(entry_obj, &["matcher"]);
            let mut origin = base_origin.clone();
            origin.event_index = event_index;
            origin.matcher_group_index = matcher_group_index;
            origin.hook_index = 0;
            hooks.push(build_hook(
                event,
                matcher,
                entry,
                origin,
                group_extra,
                path,
            )?);
        }
    }

    Ok(HookBundle { hooks })
}

fn invalid(path: &Path, message: impl Into<String>) -> AgentpackError {
    AgentpackError::io(path, io::Error::new(ErrorKind::InvalidData, message.into()))
}

fn object_extra(object: &Map<String, Value>, excluded: &[&str]) -> BTreeMap<String, Value> {
    object
        .iter()
        .filter(|(key, _)| !excluded.iter().any(|candidate| candidate == &key.as_str()))
        .map(|(key, value)| (key.clone(), value.clone()))
        .collect()
}

fn parse_handler(path: &Path, value: &Value) -> Result<(ClaudeHandler, BTreeMap<String, Value>)> {
    let object = value
        .as_object()
        .ok_or_else(|| invalid(path, "hook entries must be JSON objects"))?;
    let hook_type = object
        .get("type")
        .and_then(Value::as_str)
        .ok_or_else(|| invalid(path, "hook entries must include type"))?;
    let handler = match hook_type {
        "command" => {
            let command = object
                .get("command")
                .and_then(Value::as_str)
                .ok_or_else(|| invalid(path, "command hook missing command"))?;
            let timeout_secs = object.get("timeout").and_then(Value::as_u64);
            ClaudeHandler::Command(CommandHandler {
                command: command.to_string(),
                timeout_secs,
            })
        }
        "http" => {
            let url = object
                .get("url")
                .and_then(Value::as_str)
                .ok_or_else(|| invalid(path, "http hook missing url"))?;
            let method = object
                .get("method")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned);
            let headers = object
                .get("headers")
                .and_then(Value::as_object)
                .map(|headers| {
                    headers
                        .iter()
                        .filter_map(|(key, value)| {
                            value.as_str().map(|value| (key.clone(), value.to_string()))
                        })
                        .collect::<BTreeMap<_, _>>()
                })
                .unwrap_or_default();
            ClaudeHandler::Http(HttpHandler {
                url: url.to_string(),
                method,
                headers,
                body: object.get("body").cloned(),
            })
        }
        "prompt" => {
            let prompt = object
                .get("prompt")
                .and_then(Value::as_str)
                .ok_or_else(|| invalid(path, "prompt hook missing prompt"))?;
            ClaudeHandler::Prompt(PromptHandler {
                prompt: prompt.to_string(),
                model: object
                    .get("model")
                    .and_then(Value::as_str)
                    .map(ToOwned::to_owned),
            })
        }
        "agent" => {
            let prompt = object
                .get("prompt")
                .or_else(|| object.get("instruction"))
                .and_then(Value::as_str)
                .ok_or_else(|| invalid(path, "agent hook missing prompt"))?;
            ClaudeHandler::Agent(AgentHandler {
                prompt: prompt.to_string(),
                agent: object
                    .get("agent")
                    .and_then(Value::as_str)
                    .map(ToOwned::to_owned),
                model: object
                    .get("model")
                    .and_then(Value::as_str)
                    .map(ToOwned::to_owned),
            })
        }
        other => {
            return Err(invalid(
                path,
                format!("unsupported Claude hook type `{other}`"),
            ))
        }
    };
    let extra = match hook_type {
        "command" => object_extra(object, &["type", "command", "timeout"]),
        "http" => object_extra(object, &["type", "url", "method", "headers", "body"]),
        "prompt" => object_extra(object, &["type", "prompt", "model"]),
        "agent" => object_extra(object, &["type", "prompt", "instruction", "agent", "model"]),
        _ => BTreeMap::new(),
    };
    Ok((handler, extra))
}

fn build_hook(
    event: ClaudeEvent,
    matcher: Option<String>,
    handler_value: &Value,
    origin: HookOrigin,
    matcher_group_extra: BTreeMap<String, Value>,
    path: &Path,
) -> Result<NormalizedHook> {
    let (handler, raw_extra) = parse_handler(path, handler_value)?;
    Ok(NormalizedHook {
        event,
        matcher,
        handler,
        origin,
        matcher_group_extra,
        raw_extra,
    })
}

File: internal/hooks/parse.go
package hooks

import (
	"fmt"
	"sort"
)

func ParseNested(sourcePath string, value any, base Origin) (Bundle, error) {
	root, ok := asObject(value)
	if !ok {
		return Bundle{}, invalidHook(sourcePath, "hook file must be a JSON object")
	}
	hooks, ok := asObject(root["hooks"])
	if !ok {
		return Bundle{}, invalidHook(sourcePath, "hook file must contain a top-level hooks object")
	}
	return parseEventGroups(sourcePath, hooks, base, false)
}

func ParseCodex(sourcePath string, value any, base Origin) (Bundle, error) {
	root, ok := asObject(value)
	if !ok {
		return Bundle{}, invalidHook(sourcePath, "Codex hooks file must be an object")
	}
	if nested, nestedOK := asObject(root["hooks"]); nestedOK {
		root = nested
	}
	return parseEventGroups(sourcePath, root, base, true)
}

func parseEventGroups(sourcePath string, root map[string]any, base Origin, codex bool) (Bundle, error) {
	var bundle Bundle
	eventNames := sortedKeys(root)
	for eventIndex, eventName := range eventNames {
		event, found := ParseEvent(eventName)
		if !found {
			return Bundle{}, invalidHook(sourcePath, "unknown hook event %q", eventName)
		}
		entries, ok := root[eventName].([]any)
		if !ok {
			return Bundle{}, invalidHook(sourcePath, "event groups must be arrays")
		}
		for groupIndex, rawEntry := range entries {
			entry, ok := asObject(rawEntry)
			if !ok {
				return Bundle{}, invalidHook(sourcePath, "matcher groups must be objects")
			}
			matcher, _ := entry["matcher"].(string)
			if handlers, grouped := entry["hooks"].([]any); grouped {
				extra := objectExtra(entry, "matcher", "hooks")
				for hookIndex, rawHandler := range handlers {
					hook, err := buildHook(sourcePath, event, matcher, rawHandler, indexedOrigin(base, eventIndex, groupIndex, hookIndex), extra)
					if err != nil {
						return Bundle{}, err
					}
					bundle.Hooks = append(bundle.Hooks, hook)
				}
				continue
			}
			if !codex {
				return Bundle{}, invalidHook(sourcePath, "matcher groups must contain hooks arrays")
			}
			hook, err := buildHook(sourcePath, event, matcher, rawEntry, indexedOrigin(base, eventIndex, groupIndex, 0), objectExtra(entry, "matcher"))
			if err != nil {
				return Bundle{}, err
			}
			bundle.Hooks = append(bundle.Hooks, hook)
		}
	}
	return bundle, nil
}

func buildHook(sourcePath string, event Event, matcher string, raw any, origin Origin, groupExtra map[string]any) (Hook, error) {
	object, ok := asObject(raw)
	if !ok {
		return Hook{}, invalidHook(sourcePath, "hook entries must be JSON objects")
	}
	kindText, ok := object["type"].(string)
	if !ok {
		return Hook{}, invalidHook(sourcePath, "hook entries must include type")
	}
	handler := Handler{Kind: HandlerKind(kindText)}
	var excluded []string
	switch handler.Kind {
	case CommandHandler:
		handler.Command, ok = object["command"].(string)
		if !ok {
			return Hook{}, invalidHook(sourcePath, "command hook missing command")
		}
		if timeout, ok := unsigned(object["timeout"]); ok {
			handler.Timeout = &timeout
		}
		excluded = []string{"type", "command", "timeout"}
	case HTTPHandler:
		handler.URL, ok = object["url"].(string)
		if !ok {
			return Hook{}, invalidHook(sourcePath, "http hook missing url")
		}
		handler.Method, _ = object["method"].(string)
		handler.Headers = stringMap(object["headers"])
		handler.Body = object["body"]
		excluded = []string{"type", "url", "method", "headers", "body"}
	case PromptHandler:
		handler.Prompt, ok = object["prompt"].(string)
		if !ok {
			return Hook{}, invalidHook(sourcePath, "prompt hook missing prompt")
		}
		handler.Model, _ = object["model"].(string)
		excluded = []string{"type", "prompt", "model"}
	case AgentHandler:
		handler.Prompt, ok = object["prompt"].(string)
		if !ok {
			handler.Prompt, ok = object["instruction"].(string)
		}
		if !ok {
			return Hook{}, invalidHook(sourcePath, "agent hook missing prompt")
		}
		handler.Agent, _ = object["agent"].(string)
		handler.Model, _ = object["model"].(string)
		excluded = []string{"type", "prompt", "instruction", "agent", "model"}
	default:
		return Hook{}, invalidHook(sourcePath, "unsupported Claude hook type %q", kindText)
	}
	return Hook{Event: event, Matcher: matcher, Handler: handler, Origin: origin, MatcherGroupExtra: groupExtra, RawExtra: objectExtra(object, excluded...)}, nil
}

func indexedOrigin(origin Origin, event, group, hook int) Origin {
	origin.EventIndex, origin.MatcherGroupIndex, origin.HookIndex = event, group, hook
	return origin
}

func asObject(value any) (map[string]any, bool) {
	object, ok := value.(map[string]any)
	return object, ok
}

func sortedKeys(object map[string]any) []string {
	keys := make([]string, 0, len(object))
	for key := range object {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

func objectExtra(object map[string]any, excluded ...string) map[string]any {
	skip := make(map[string]bool, len(excluded))
	for _, key := range excluded {
		skip[key] = true
	}
	extra := make(map[string]any)
	for key, value := range object {
		if !skip[key] {
			extra[key] = value
		}
	}
	return extra
}

func unsigned(value any) (uint64, bool) {
	switch number := value.(type) {
	case float64:
		if number >= 0 && number == float64(uint64(number)) {
			return uint64(number), true
		}
	case uint64:
		return number, true
	case int:
		if number >= 0 {
			return uint64(number), true
		}
	}
	return 0, false
}

func stringMap(value any) map[string]string {
	object, _ := asObject(value)
	result := make(map[string]string)
	for key, value := range object {
		if text, ok := value.(string); ok {
			result[key] = text
		}
	}
	return result
}

func invalidHook(path, format string, arguments ...any) error {
	return fmt.Errorf("invalid hook file %s: %s", path, fmt.Sprintf(format, arguments...))
}
